lib/workload/stateless/stacks/icav2-event-translator/translator_service/icav2_event_translator.py

```diff
@@ -0,0 +1,251 @@
+      dynamodb.update_item(
+          TableName=table_name,
+          Key={
+              'id': {'S': internal_ica_event.detail.payload.get("analysisId", '')},
+              'id_type': {'S': 'analysis_id'}
+          },
+          UpdateExpression='SET db_uuid = :db_uuid',
+          ExpressionAttributeValues={
+              ':db_uuid': {'S': db_uuid}
+          }
+      )
+      dynamodb.update_item(
+          TableName=table_name,
+          Key={
+              'id': {'S': internal_ica_event.detail.portalRunId},
+              'id_type': {'S': 'portal_run_id'}
+          },
+          UpdateExpression='SET db_uuid = :db_uuid',
+          ExpressionAttributeValues={
+              ':db_uuid': {'S': db_uuid}
+          }
+      )
+      logger.info(f"db_uuid updated in anaylsis_id and portal_run_id record.")
+    except Exception as e:
+        raise Exception("Failed to store event in the DynamoDB table. Error: ", e)
+
+    return {
+        "statusCode": 200,
+        "body": json.dumps("Internal event sent to the event bus and both msg stored in the DynamoDB table.")
+    }
+
+# Convert from Entity model to aws event object with aws event envelope
+def translate_to_aws_event(event)->AWSEvent:
+  return AWSEvent(
+    detail= get_event_details(event),
+    detail_type= "WorkflowRunStateChange",
+    source= "orcabus.bcm",
+    # version="0.1.0",  # comment as the version is managed by the evnet bus
+  )
+
+# Convert from entity module to internal event details
+def get_event_details(event)->WorkflowRunStateChange:
+    # Extract relevant fields from the event payload
+    project_id = event.get("projectId", '')
+    analysis_status = event.get("eventParameters", {}).get("analysisStatus", '')
+    
+    payload = event.get("payload", {})
+    pipeline = payload.get("pipeline", {})
+    
+    # generate internal event with required attributes
+    return WorkflowRunStateChange(
+      portal_run_id= get_portal_run_id(payload.get("id", '')),
+      timestamp= datetime.datetime.now(datetime.timezone.utc).strftime('%Y-%m-%dT%H:%M:%SZ'),
+      status= analysis_status,
+      workflow_type= "bssh_bcl_convert",
+      workflow_version= "4.2.7",
+      payload= {
+        "refId": None,
+        "version": "0.1.0",
+        "projectId": project_id,
+        "analysisId": payload.get("id", ''),
+        "userReference": payload.get("userReference", ''),
+        "timeCreated": payload.get("timeCreated",""),
+        "timeModified": payload.get("timeModified",""),
+        "pipelineId": pipeline.get("id",''),
+        "pipelineCode": pipeline.get("code",''),
+        "pipelineDescription": pipeline.get("description",''),
+        "pipelineUrn": pipeline.get("urn",'')
+      }
+    )
+
+# check the dynamodb table to see if new portal run id is required
+def get_portal_run_id(analysis_id: str) -> str:
+    
+    # check if dynomodb table have the anylsis id
+    table_name = os.getenv("TABLE_NAME")
+    try:
+      response = dynamodb.scan(
```

Comment:
Hmm, I suppose the `id` and `id_type` is the PK and SK, so maybe use `query` instead of `scan`? `scan` (scan) the whole table where query narrows down to that PK and SK then scan for the filter attribute.

I guess this might be a problem when the table gets bigger.

Ref:
- https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/bp-query-scan.html#bp-query-scan-performance
- https://dynobase.dev/dynamodb-scan-vs-query/

Reply:
Yup, would be good if we can leverage `query` lookup with dynamodb whenever possible.

Reply:
sure, will update this in next commit

Reply:
fixed, definitely query is better than scan here, thanks.